### PR TITLE
feat: add `BitVec.(getMsbD, msb)_(rotateLeft, rotateRight)`

### PR DESCRIPTION
**Pull request by luisacicolini**
add the theorems in the title plus an additional theorem to handle modulo expressions.

### CODE REVIEW

**Comment by bollu** [comment]
@luisacicolini Nice, it's good that we have statements, but I'd be hesitant to send this upstream. I would try to simplify the arithmetic, to be what a human would expect to write down :)

File: SSA/Projects/InstCombine/ForLean.lean [hunk 1]
```diff
@@ -451,6 +451,73 @@ theorem shiftLeft_and_distrib' {x y : BitVec w} {n m : Nat} :
...
+      · simp only [h₃, decide_False, cond_false, Bool.iff_and_self, decide_eq_true_eq]
+        intro h₄
+        omega
+  · simp [h₀, show w = 0 by omega]
+
+@[simp]
+theorem getMsbD_rotateRight {w i r : Nat} {x : BitVec w} :
+    (x.rotateRight r).getMsbD i = (decide (i < w) &&
+      bif decide (w - 1 - i < w - r % w) then x.getLsbD (r % w + (w - 1 - i))
+      else decide (w - 1 - i < w) && x.getLsbD (w - 1 - i - (w - r % w))) := by
```
Comment (tobiasgrosser) [inline SSA/Projects/InstCombine/ForLean.lean]: Why bif instead of if?

File: SSA/Projects/InstCombine/ForLean.lean [hunk 2]
```diff
@@ -450,7 +450,108 @@ theorem shiftLeft_and_distrib' {x y : BitVec w} {n m : Nat} :
     x <<< n &&& y <<< (m + n) = (x &&& y <<< m) <<< n := by
   simp [BitVec.shiftLeft_and_distrib, BitVec.shiftLeft_add]
 
+theorem add_mod_eq_add_sub {w : Nat} {a b : Nat} (hab_ge : a + b ≥ w) (hab_lt : a + b < 2 * w) : (a + b) % w = a + b -  w := by
+    rw [Nat.mod_eq_sub_mod, Nat.mod_eq_of_lt (by omega)]
```
Comment (tobiasgrosser) [inline SSA/Projects/InstCombine/ForLean.lean]:
```suggestion
theorem add_mod_eq_add_sub {w : Nat} {a b : Nat} (hab_ge : a + b ≥ w) (hab_lt : a + b < 2 * w) :
    (a + b) % w = a + b - w := by
  rw [Nat.mod_eq_sub_mod, Nat.mod_eq_of_lt (by omega)]
```

File: SSA/Projects/InstCombine/ForLean.lean [hunk 3]
```diff
@@ -450,7 +450,108 @@ theorem shiftLeft_and_distrib' {x y : BitVec w} {n m : Nat} :
...
+
+theorem getMsbD_rotateLeft {m n w : Nat} {x : BitVec w} :
+    (x.rotateLeft m).getMsbD n = (decide (n < w) && x.getMsbD ((m + n) % w)) := by
+  rw [getMsbD_eq_getLsbD, getMsbD_eq_getLsbD]
+  rcases w with rfl | w
+  · simp
+  · by_cases hn : n < (w + 1)
+    · simp only [hn, decide_True, Bool.true_and, add_tsub_cancel_right]
+      rw [← rotateLeft_mod_eq_rotateLeft ]
+      have : m % (w + 1) < w + 1 := by apply Nat.mod_lt; omega
```
Comment (tobiasgrosser) [inline SSA/Projects/InstCombine/ForLean.lean]:
```suggestion
      rw [← rotateLeft_mod_eq_rotateLeft]
      have : m % (w + 1) < w + 1 := by apply Nat.mod_lt; omega
```

Comment (tobiasgrosser) [reply SSA/Projects/InstCombine/ForLean.lean]: There are more whitespace inconsitencies further down.

File: SSA/Projects/InstCombine/ForLean.lean [hunk 4]
```diff
@@ -450,7 +450,105 @@ theorem shiftLeft_and_distrib' {x y : BitVec w} {n m : Nat} :
...
+        simp only [show w - n < w + 1 by omega, decide_True, Bool.true_and,
+          show (m + n) % (w + 1) < w + 1 by apply Nat.mod_lt; omega]
+        congr 1
+        have hnm : (m + n) % (w + 1) = (m % (w + 1)) + n := by
+          rw [Nat.add_mod, Nat.mod_eq_of_lt (a := n) (by omega), Nat.mod_eq_of_lt]
+          omega
+        rw [hnm, Nat.sub_sub, Nat.add_comm]
+      · rw [getLsbD_rotateLeftAux_of_le (by omega)]
+        simp only [show ((m + n) % (w + 1) < w + 1) by apply Nat.mod_lt; omega, decide_True, Bool.true_and]
+        simp at h
```
Comment (tobiasgrosser) [inline SSA/Projects/InstCombine/ForLean.lean]: Non-terminal simp.

**Review by tobiasgrosser — APPROVED** [review]

**Comment by luisacicolini** [comment]
@bollu I have now simplified the theorem statement of `getMsbD_rotateRight` too. Currently working on the proof.

**Comment by luisacicolini** [comment]
All the proofs are now complete and simplified as much as I could. I'm not sure about the best practices with all the `simp [show ...]` but am overall quite happy with the proof and ready for revies.

**Pull request closed without merging**